models/surveys.go

```diff
@@ -29,53 +126,177 @@ func AllSurveys() ([]*SurveySummary, error) {
 	if err != nil {
-		return nil, err
+		if err == sql.ErrNoRows {
+			http.Error(w, "Survey not found", http.StatusNotFound)
+		} else {
+			http.Error(w, "get survey by reference query failed", http.StatusInternalServerError)
+		}
+	}
+
+	data, _ := json.Marshal(survey)
+	w.Write(data)
+
+}
+
+// AllClassifierTypeSelectors returns all the classifier type selectors for the survey identified by the string surveyID. The classifier type selectors are returned in ascending order.
+func (api *API) AllClassifierTypeSelectors(w http.ResponseWriter, r *http.Request) {
+	// We need to run a query first to check if the survey exists so an HTTP 404 can be correctly
+	// returned if it doesn't exist. Without this check an HTTP 204 is incorrectly returned for an
+	// invalid survey ID.
+	vars := mux.Vars(r)
+	surveyId := vars["surveyId"]
+
+	err := api.getSurveyID(surveyId)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			http.Error(w, "Survey not found", http.StatusNotFound)
+		} else {
+			http.Error(w, "Error getting list of classifier type selectors for survey '"+surveyId+"' - "+err.Error(), http.StatusInternalServerError)
+		}
+	}
+
+	// Now we can get the classifier type selector records.
+	rows, err := api.GetClassifierTypeSelectorStmt.Query(surveyId)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			http.Error(w, "Classifier type selector not found", http.StatusNotFound)
+		} else {
+			http.Error(w, "Error getting list of classifier type selectors for survey '"+surveyId+"' - "+err.Error(), http.StatusInternalServerError)
+		}
+	}
+
+	defer rows.Close()
+	classifierTypeSelectorSummaries := make([]*ClassifierTypeSelectorSummary, 0)
+
+	for rows.Next() {
+		classifierTypeSelectorSummary := new(ClassifierTypeSelectorSummary)
+		err := rows.Scan(&classifierTypeSelectorSummary.ID, &classifierTypeSelectorSummary.Name)
+
+		if err != nil {
+			//LogError("Error getting list of classifier type selectors for survey '"+surveyID+"'", err)
+			http.Error(w, "Error getting list of classifier type selectors for survey '"+surveyId+"' - "+err.Error(), http.StatusInternalServerError)
+		}
+
+		classifierTypeSelectorSummaries = append(classifierTypeSelectorSummaries, classifierTypeSelectorSummary)
 	}
 
-	return survey, nil
+	data, _ := json.Marshal(classifierTypeSelectorSummaries)
+	w.Write(data)
 }
 
-func getSurveyID(surveyID string) error {
+// GetClassifierTypeSelector returns the details of the classifier type selector for the survey identified by the string surveyID and
+// the classifier type selector identified by the str	ing classifierTypeSelectorID.
+func (api *API) GetClassifierTypeSelectorById(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	surveyId := vars["surveyId"]
+	classifierTypeSelectorId := vars["classifierTypeSelectorId"]
+
+	err := api.getSurveyID(surveyId)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			http.Error(w, "Survey not found", http.StatusNotFound)
```

Comment:
The previous version of the survey service had a "RESTError" model to represent a 404 not found - i think other services calling expect this data structure so we need to use it for the time being